tests/backend/checkmach/test_assume_error.ml

```diff
@@ -0,0 +1,225 @@
+    | _ ->
+      Printf.printf "BOO %x" x;  (* ignore this indirect call *)
+      (x,0)
+end
+
+(* A3: shows that "assume error" causes the check to ignore allocations
+   in [foo] that come after the exceptional return from [f x]
+   and correctly report allocations that happen after normal return
+   from [f x]. *)
+module A3 = struct
+  let[@inline never][@local never] h s = [Random.int s ; Random.int s]
+
+  let[@inline never][@local never][@zero_alloc] g x =
+    if x < 0 then Sys.opaque_identity (x+1) else raise (Failure ((string_of_int) x))
+
+  let[@inline never][@local never] handle_exn exn =
+    (match exn with
+     | Failure s -> print_string s; print_newline ()
+     | _ -> failwith "Boo");
+    0
+  ;;
+
+  let[@zero_alloc] f x =
+    match g x with
+    | exception exn ->
+      (handle_exn[@zero_alloc assume error]) exn
+      |> h |> List.hd  (* ignore this allocation *)
+    | answer -> (answer * 2)
+
+  let[@zero_alloc] foo x =
+    try
+      let y = f x in
+      (x,y)  (* report this allocation *)
+    with
+    | _ ->
+      Printf.printf "BOO %x" x;  (* ignore this indirect call *)
+      (x,0)
+end
+
+(* A4: is a succesful version of A3, without allocation after normal return
+   from [f x]. *)
+module A4 = struct
+  let[@inline never][@local never] h s = [Random.int s ; Random.int s]
+
+  let[@inline never][@local never][@zero_alloc] g x =
+    if x < 0 then Sys.opaque_identity (x+1) else raise (Failure ((string_of_int) x))
+
+  let[@inline never][@local never] handle_exn exn =
+    (match exn with
+     | Failure s -> print_string s; print_newline ()
+     | _ -> failwith "Boo");
+    0
+  ;;
+
+  let[@zero_alloc] f x =
+    match g x with
+    | exception exn ->
+      (handle_exn[@zero_alloc assume error]) exn
+      |> h |> List.hd  (* ignore this allocation *)
+    | answer -> (answer * 2)
+
+  let[@zero_alloc] foo x =
+    try f x
+    with
+    | _ ->
+      Printf.printf "BOO %x" x;  (* ignore this indirect call *)
+      0
+end
+
+(* A5: same as A4 but with "assume error" on the function definition instead
+   of the call site. *)
+module A5 = struct
+  let[@inline never][@local never] h s = [Random.int s ; Random.int s]
+
+  let[@inline never][@local never][@zero_alloc] g x =
+    if x < 0 then Sys.opaque_identity (x+1) else raise (Failure ((string_of_int) x))
+
+  let[@inline never][@local never][@zero_alloc assume error] handle_exn exn =
+    (match exn with
+     | Failure s -> print_string s; print_newline ()
+     | _ -> failwith "Boo");
+    0
+  ;;
+
+  let[@zero_alloc] f x =
+    match g x with
+    | exception exn ->
+      handle_exn exn
+      |> h |> List.hd  (* ignore this allocation *)
+    | answer -> (answer * 2)
+
+  let[@zero_alloc] foo x =
+    try f x
+    with
+    | _ ->
+      Printf.printf "BOO %x" x;  (* ignore this indirect call *)
+      0
+end
+
+(* A6: shows that "assume error" is not enough to prove "strict",
+   because the call to [string_of_int] is not known to not diverge.
```

Comment:
Why would knowing `string_of_int` doesn't diverge help us here?  Aren't we still going to end up allocating the string on the path to `handle_exn`, which should be disallowed by the strict check?

Reply:
The analysis thinks that `string_of_int` can diverge, so  it can't prove that `f` is "zero_alloc strict". I'm trying to show that `[@zero_alloc assume error]` doesn't affect the `div`component of the abstract value. It doesn't matter that we allocate on the path to `handle_exn` because "assume error" causes all allocations before and after it to be ignored, as if we had "exit 1" inside `handle_exn`, rather than a `raise`. (In the analysis is implemented by "assume error" does not return and does not raise.)

Reply:
Thanks.  To check my understanding, the error say "may allocate on a path to diverge".  Is the potential allocation also within string_of_int? I don't see another.

Reply:
yes, the allocation is within `string_of_int`. I updated the comment on the test.